File: Prácticas/P1/Ruby/civitas/lib/mazo_sorpresas.rb
# encoding: utf-8
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "diario.rb"
require_relative "sorpresa.rb"

module Civitas
  class MazoSorpresas
    def initialize(debug = false)
      init
      @debug = debug
      @ultima_sorpresa
      @diario = Diario.instance
      if @debug == true
        @diario.ocurre_evento("MazoSorpresas en modo debug")
      end
    end
    #private_class_method :new
    
    private
    def init
      @sorpresas = Array.new
      @barajada = false
      @usadas = 0
      @cartas_especiales = Array.new
    end
    
    public
    def al_mazo(s)
      if(@barajada == false)
        @sorpresas << s
      end
    end
    
    def siguiente
      if @barajada == false && @usadas == @sorpresas.length && @debug == false
        @sorpresas.shuffle
        @usadas = 0;
        @barajada = true;
      end
      @usadas = @usadas + 1
      @ultima_sorpresa = @sorpresas.first
      @sorpresas.shift
      @sorpresas << @ultima_sorpresa
      @ultima_sorepresa
    end
    
    def inhabilitar_carta_especial(sorpresa)
      if @sorpresas.index(sorpresa) != nil
        @cartas_especiales << @sorpresas[@sorpresas.index(sorpresa)]
        @sorpresas.delete(sorpresa)
        @diario.ocurre_evento("Carta especial inhabilitada")
      end
    end
    
    def habilitar_carta_especial(sorpresa)
      if @cartas_especiales.index(sorpresa) != nil
        @sorpresas << @cartas_especiales[@cartas_especiales.index(sorpresa)]
        @cartas_especiales.delete(sorpresa)
        @diario.ocurre_evento("Carta especial habilitada")
      end
    end
  end
end

File: Prácticas/P2/Java/civitas/src/main/java/civitas/MazoSorpresas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Locale;

/**
 *
 * @author Jose Padial
 */
public class MazoSorpresas {
    private ArrayList<Sorpresa> sorpresas;
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;
    private Diario diario = Diario.getInstance();
    
    private void init(){
        sorpresas = new ArrayList();
        cartasEspeciales = new ArrayList();
        barajada = false;
        usadas = 0;
    }

    MazoSorpresas(){
        init();
        debug = false;
    }
    
    MazoSorpresas(boolean debug){
        this.debug = debug;
        init();
        if(this.debug == true)
            diario.ocurreEvento("MazoSorpresas en modo debug");
    }
    
    void alMazo(Sorpresa s){
        if(barajada == false)
            sorpresas.add(s);
    }
    
    Sorpresa siguiente(){
        if(barajada == false || usadas == sorpresas.size() || debug == false){
            Collections.shuffle(sorpresas);
            usadas = 0;
            barajada = true;
        }
        usadas++;
        ultimaSorpresa = sorpresas.get(0);
        sorpresas.remove(0);
        sorpresas.add(ultimaSorpresa);
        return ultimaSorpresa;
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        if(sorpresas.indexOf(sorpresa) != -1){
            cartasEspeciales.add(sorpresas.get(sorpresas.indexOf(sorpresa)));
            sorpresas.remove(sorpresa);
            diario.ocurreEvento("Carta especial inhabilitada");
        }
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
        if(cartasEspeciales.indexOf(sorpresa) != -1){
            sorpresas.add(cartasEspeciales.get(cartasEspeciales.indexOf(sorpresa)));
            cartasEspeciales.remove(sorpresa);
            diario.ocurreEvento("Carta especial habilitada");  
        }
    }

    Sorpresa getUltimaSorpresa() {
        return ultimaSorpresa;
    }
        
}
